File: tools/wasmtime-spike/android-harness/app/src/main/java/org/akkirrai/beakokit/runtime/NativeSourceRuntimeBridge.java
package org.akkirrai.beakokit.runtime;

public final class NativeSourceRuntimeBridge {
    private NativeSourceRuntimeBridge() {}

    public interface Host {
        byte[] call(byte[] request);
    }

    public static native String protocolModuleCallWithHost(
            byte[] module,
            String request,
            Host host,
            long cancellationScopeId);
}

File: shared/src/androidMain/kotlin/org/akkirrai/beakokit/runtime/NativeSourceRuntimeBridge.kt
package org.akkirrai.beakokit.runtime

/** Android entry point for the packaged Wasmtime runtime. */
class NativeSourceRuntimeBridge private constructor() {
    fun interface Host {
        fun call(request: ByteArray): ByteArray
    }

    companion object {
        init {
            System.loadLibrary("wasmtime_spike")
        }

        @JvmStatic
        external fun validateModule(module: ByteArray)

        @JvmStatic
        external fun protocolModuleCallWithHost(
            module: ByteArray,
            request: String,
            host: Host,
            cancellationScopeId: Long,
        ): String

        @JvmStatic
        external fun beginCancellationScope(): Long

        @JvmStatic
        external fun cancelCancellationScope(scopeId: Long)

        @JvmStatic
        external fun finishCancellationScope(scopeId: Long)
    }
}
